### PR TITLE
chore: turn on clippy::pedantic

### PR DESCRIPTION
**Pull request by mxinden**
Enable clippy's _pedantic_ lint category.

- Adjusts CI check accordingly.
- Adds `#![deny(clippy::pedantic)]` to all (but tests) top level crate files.
- Fixes various lint failures.
- Removes module `#![deny(clippy::pedantic)]` whereever it is already enabled at the crate level.

Fixes https://github.com/mozilla/neqo/issues/553

More work needed here. _Draft_ for now.

### CODE REVIEW

File: neqo-client/src/main.rs [hunk 1]
```diff
@@ -6,6 +6,7 @@
 
 #![cfg_attr(feature = "deny-warnings", deny(warnings))]
 #![warn(clippy::use_self)]
+#![deny(clippy::pedantic)]
```
Comment (martinthomson) [inline neqo-client/src/main.rs:9]: In other projects, we use `warn(clippy::pedantic)` in code, but then execute pre-commit hooks and CI with `-D warnings`.  I think that is a better approach overall than having this in code.

File: neqo-transport/src/addr_valid.rs [hunk 2]
```diff
@@ -234,16 +234,16 @@ impl AddressValidation {
                     panic!("AddressValidation: Retry token with small CID {}", cid);
                 }
             } else if cid.is_empty() {
-                // An empty connection ID means NEW_TOKEN.
+                // An empty connection ID means `NEW_TOKEN`.
                 if self.validation == ValidateAddress::Always {
-                    qinfo!("AddressValidation: valid NEW_TOKEN token; validating again");
+                    qinfo!("AddressValidation: valid `NEW_TOKEN` token; validating again");
```
Comment (martinthomson) [inline neqo-transport/src/addr_valid.rs:239]: I don't think that we need these to be quoted like this.

File: neqo-transport/src/lib.rs [hunk 3]
```diff
@@ -6,6 +6,18 @@
...
+#![allow(clippy::must_use_candidate)]
+#![allow(clippy::return_self_not_must_use)]
+#![allow(clippy::module_name_repetitions)]
+#![allow(clippy::missing_errors_doc)]
+#![allow(clippy::needless_pass_by_value)]
+#![allow(clippy::doc_markdown)]
+#![allow(clippy::too_many_lines)]
+#![allow(clippy::single_match_else)]
+#![allow(clippy::if_not_else)]
+#![allow(clippy::match_wild_err_arm)]
```
Comment (martinthomson) [inline neqo-transport/src/lib.rs:20]: These are fairly broad and I'm not sure that I like a blanket allow on each of them.  (There are some that might be justified, for sure, but not all of these.)

**Review by martinthomson — COMMENTED** [review]
Making good progress. :)

**Comment by mxinden** [comment]
Thank you for the review!

Closing here. I will create individual pull requests (see e.g. linked ones above), to ease review and to reduce merge conflicts. Still lots to do.

**Pull request closed without merging**